File: FE/macro.py
#!/bin/env python3

"""Macro Expander

"""

import dataclasses
import logging

from typing import Any, Tuple, Union

from FE import cwast
from FE import identifier
from FE import symbolize

logger = logging.getLogger(__name__)


class _MacroContext:
    """TBD"""

    def __init__(self, id_gen: identifier.IdGen, srcloc):
        self._id_gen = id_gen
        # these need to become lists
        self.macro_parameter: dict[cwast.NAME, Any] = {}
        self.srcloc = srcloc

    def SetSymbol(self, name: cwast.NAME, value):
        assert name.IsMacroVar()
        self.macro_parameter[name] = value

    def RegisterSymbol(self, name: cwast.NAME, value):
        assert name not in self.macro_parameter, f"duplicate macro param: {name} in {value}"
        self.SetSymbol(name, value)

    def GetSymbol(self: Any, name: cwast.NAME) -> Any:
        assert name.IsMacroVar()
        return self.macro_parameter[name]

    def GenerateNewSymbol(self, name: cwast.NAME, srcloc):
        # print (f"@@@@@@@@@@@@@@@ {name}")
        new_name = self._id_gen.NewName(name.name[1:])
        self.RegisterSymbol(name, cwast.Id(
            new_name, None, x_srcloc=srcloc))


def _ExpandMacroBodyNodeRecursively(node, ctx: _MacroContext) -> Any:
    if isinstance(node, cwast.DefVar) and node.name.IsMacroVar():
        new_name = ctx.GetSymbol(node.name)
        # Why is this not a MacroVar
        assert isinstance(new_name, cwast.Id)
        type_or_auto = _ExpandMacroBodyNodeRecursively(node.type_or_auto, ctx)
        initial = _ExpandMacroBodyNodeRecursively(
            node.initial_or_undef_or_auto, ctx)
        return cwast.DefVar(new_name.GetBaseNameStrict(), type_or_auto, initial,
                            x_srcloc=ctx.srcloc, mut=node.mut, ref=node.ref)
    elif isinstance(node, cwast.MacroId):
        arg = ctx.GetSymbol(node.name)
        replacement = cwast.CloneNodeRecursively(arg, {}, {})
        # when we expland the MacroId we unwrap the EphemeralList
        if isinstance(replacement, cwast.EphemeralList):
            return replacement.args
        return replacement
    elif isinstance(node, cwast.MacroFor):
        loop_var = node.name
        assert loop_var.IsMacroVar(), f" non macro name: {node}"
        args = ctx.GetSymbol(node.name_list)
        assert isinstance(args, cwast.EphemeralList)
        out = []
        for item in args.args:
            ctx.SetSymbol(loop_var, item)
            for b in node.body_for:
                exp = _ExpandMacroBodyNodeRecursively(b, ctx)
                if isinstance(exp, list):
                    out += exp
                else:
                    out.append(exp)
        return out

    clone = dataclasses.replace(node)

    for nfd in node.__class__.NODE_FIELDS:
        c = nfd.name
        if nfd.kind is cwast.NFK.NODE:
            replacement = _ExpandMacroBodyNodeRecursively(
                getattr(node, c), ctx)
            assert not isinstance(replacement, list)
            setattr(clone, c, replacement)
        else:
            out = []
            for cc in getattr(node, c):
                exp = _ExpandMacroBodyNodeRecursively(cc, ctx)
                # TODO: this tricky and needs a comment
                if isinstance(exp, list):
                    out += exp
                else:
                    out.append(exp)
            setattr(clone, c, out)
    return clone


def _CheckMacroArg(p,  a, macro_invoke, def_macro):
    assert p.name.IsMacroVar()
    kind = p.macro_param_kind
    if kind is cwast.MACRO_PARAM_KIND.EXPR:
        pass
    elif kind in (cwast.MACRO_PARAM_KIND.STMT_LIST,
                  cwast.MACRO_PARAM_KIND.EXPR_LIST_REST):
        if not isinstance(a, cwast.EphemeralList):
            cwast.CompilerError(macro_invoke.x_srcloc,
                                f"expected EphemeralList for macro param {p} got {a}")
    elif kind is cwast.MACRO_PARAM_KIND.TYPE:
        pass
    elif kind is cwast.MACRO_PARAM_KIND.FIELD:
        assert isinstance(a, cwast.Id)
    elif kind in (cwast.MACRO_PARAM_KIND.ID, cwast.MACRO_PARAM_KIND.ID_DEF):
        assert isinstance(
            a, cwast.Id), f"while expanding macro {def_macro.name} expected parameter id but got: {a}"
    else:
        assert False


MAX_MACRO_NESTING = 8


def _FixUpExprListRest(params: list[cwast.MacroParam], args: list[Any]):
    """collect EXPR_LIST_REST parameters into an EphemeralList"""
    if params and params[-1].macro_param_kind is cwast.MACRO_PARAM_KIND.EXPR_LIST_REST:
        rest = cwast.EphemeralList(args[len(params)-1:], colon=False)
        del args[len(params)-1:]
        args.append(rest)


def _ExpandMacroInvocation(macro_invoke: cwast.MacroInvoke,
                           nesting, id_gen: identifier.IdGen) -> Any:
    if nesting >= MAX_MACRO_NESTING:
        cwast.CompilerError(macro_invoke.x_srcloc, "nesting to deep")
    def_macro: cwast.DefMacro = macro_invoke.x_symbol
    assert isinstance(
        def_macro, cwast.DefMacro), f"{node} -> {def_macro}"
    params: list[cwast.MacroParam] = def_macro.params_macro
    args = macro_invoke.args
    _FixUpExprListRest(params, args)
    if len(params) != len(args):
        cwast.CompilerError(macro_invoke.x_srcloc, f"parameter mismatch in: {macro_invoke}: "
                            f"actual {len(args)} expected: {len(params)}")
    logger.info("Expanding Macro Invocation: %s", macro_invoke)
    # pp_sexpr.PrettyPrint(invoke)
    # pp_sexpr.PrettyPrint(macro)
    ctx = _MacroContext(id_gen, macro_invoke.x_srcloc)
    for p, a in zip(params, args):
        _CheckMacroArg(p, a, macro_invoke, def_macro)
        ctx.RegisterSymbol(p.name, a)
    for gen_id in def_macro.gen_ids:
        assert isinstance(gen_id, cwast.MacroId)
        # new_name = cwast.NAME.Make(gen_id.name.name[1:])
        ctx.GenerateNewSymbol(gen_id.name, def_macro.x_srcloc)

    out = []
    for node in def_macro.body_macro:
        logger.debug("Expand macro body node: %s", node)
        # pp_sexpr.PrettyPrint(node)
        exp = _ExpandMacroBodyNodeRecursively(node, ctx)
        # pp_sexpr.PrettyPrint(exp)
        if isinstance(exp, list):
            out += exp
        else:
            out.append(exp)
    # using a EphemeralList is somewhat ugly but we can run
    # _ExpandMacrosAndMacroLikeRecursively more easily this way.
    x = cwast.EphemeralList(out)
    # expand the macro body
    _ExpandMacrosAndMacroLikeRecursively(x, nesting + 1, id_gen)
    return x.args


def _ExpandMacrosAndMacroLikeRecursively(fun,  nesting: int, id_gen: identifier.IdGen):
    def replacer(node: Any):
        nonlocal nesting, id_gen
        orig_node = node
        if isinstance(node, cwast.MacroInvoke):
            nodes = _ExpandMacroInvocation(node, nesting, id_gen)
            # pp_sexpr.PrettyPrint(exp)
            if len(nodes) == 1:
                return nodes[0]
            return nodes
        elif isinstance(node, cwast.ExprSrcLoc):
            return cwast.ValString(f'r"{node.expr.x_srcloc}"', x_srcloc=node.x_srcloc)
        elif isinstance(node, cwast.ExprStringify):
            # assert isinstance(node.expr, cwast.Id)
            return cwast.ValString(f'r"{node.expr}"', x_srcloc=node.x_srcloc)
        return None if orig_node is node else node

    cwast.MaybeReplaceAstRecursivelyPost(fun, replacer)


def ExpandMacrosAndMacroLike(mods: list[cwast.DefMod]):
    """Expands MacroInvoke, ExprSrcLoc, ExprStringify"""
    for mod in mods:
        for node in mod.body_mod:
            if isinstance(node, cwast.DefFun):
                logger.info("Expanding macros in: %s", node.name)
                _ExpandMacrosAndMacroLikeRecursively(
                    node, 0, identifier.IdGen())
    # now that all macro invocations have been expanded,
    # we do not need the macros anymore
    for mod in mods:
        cwast.RemoveNodesOfType(mod, cwast.DefMacro)

############################################################
#
############################################################


def main(argv: list[str]):
    assert len(argv) == 1
    fn = argv[0]
    fn, ext = os.path.splitext(fn)
    assert ext in (".cw", ".cws")
    cwd = os.getcwd()
    main = str(pathlib.Path(fn).resolve())
    mp = mod_pool.ReadModulesRecursively(pathlib.Path(
        cwd) / "Lib", [main], add_builtin=fn != "Lib/builtin")
    mod_topo_order = mp.mods_in_topo_order
    for mod in mod_topo_order:
        canonicalize.FunRemoveParentheses(mod)
    ExpandMacrosAndMacroLike(mod_topo_order)
    symbolize.SetTargetFields(mp.mods_in_topo_order)
    symbolize.ResolveSymbolsInsideFunctions(
        mod_topo_order, mp.builtin_symtab)
    for ast in mod_topo_order:
        # cwast.CheckAST(ast, set())
        symbolize.VerifySymbols(ast)
        pp_sexpr.PrettyPrint(ast, sys.stdout)


if __name__ == "__main__":
    import sys
    import os
    import pathlib
    from FE import mod_pool
    from FE import canonicalize
    from FE import pp_sexpr

    logging.basicConfig(level=logging.WARNING)
    symbolize.logger.setLevel(logging.INFO)
    logger.setLevel(logging.INFO)

    main(sys.argv[1:])

File: FE/macro.cc

#include "FE/macro.h"

#include <set>
#include <sstream>

#include "FE/cwast_gen.h"
#include "FE/symbolize.h"

namespace cwerg::fe {

class MacroContext {
 private:
  Node invoke_;
  IdGen* id_gen_;
  SrcLoc srcloc_;
  std::map<Name, Node> symtab_;
  std::map<Name, Node> symtab_not_owned_;

 public:
  MacroContext(Node invoke, IdGen* id_gen, const SrcLoc& srcloc)
      : invoke_(invoke), id_gen_(id_gen), srcloc_(srcloc) {}

  ~MacroContext() {
    for (const auto& [name, node] : symtab_) {
      if (!symtab_not_owned_.contains(name)) NodeFreeRecursively(node);
    }
  }

  void SetSymbol(Name name, Node value) {
    // std::cout << "@@ SetSymbol " << Node_name(invoke_) << " " << name << " "
    //          << EnumToString(Node_kind(value)) << " " << value.index() <<
    //          "\n";
    ASSERT(NameIsMacro(name), "");
    symtab_not_owned_[name] = value;
  }

  Node GetSymbol(Name name) {
    auto it = symtab_.find(name);
    if (it != symtab_.end()) {
      return it->second;
    }
    it = symtab_not_owned_.find(name);
    if (it != symtab_not_owned_.end()) {
      return it->second;
    }
    ASSERT(false, "unknown name " << name);
    return kNodeInvalid;
  }

  void RegisterSymbolWithOwnership(Name name, Node value) {
    // std::cout << "@@ SetSymbol WO " << Node_name(invoke_) << " " << name << "
    // "
    //          << EnumToString(Node_kind(value)) << " " << value.index() <<
    //          "\n";
    ASSERT(!symtab_.contains(name), "");
    ASSERT(NameIsMacro(name), "");
    symtab_[name] = value;
  }

  void GenerateNewSymbol(Name name, const SrcLoc& srcloc) {
    Name new_name = id_gen_->NameNewNext(NameNew(NameData(name) + 1));
    Node id = NodeNew(NT::Id);
    NodeInitId(id, new_name, kNameInvalid, kStrInvalid, srcloc);

    RegisterSymbolWithOwnership(name, id);
  }
};

Node FixUpArgsForExprListRest(Node params, Node args) {
  int num_params = NodeNumSiblings(params);
  Node last_param = NodeLastSiblings(params);
  if (Node_macro_param_kind(last_param) != MACRO_PARAM_KIND::EXPR_LIST_REST ||
      args.isnull()) {
    return args;
  }

  Node rest = NodeNew(NT::EphemeralList);
  if (num_params == 1) {
    NodeInitEphemeralList(rest, args, 0, kStrInvalid, Node_srcloc(args));
    return rest;
  }

  Node head = args;
  // advance args to the one before the "rest"
  for (int i = 0; i < num_params - 2; ++i) {
    if (args.isnull()) {
      CompilerError(Node_srcloc(args)) << "too few arguments";
    }
    args = Node_next(args);
  }

  NodeInitEphemeralList(rest, Node_next(args), 0, kStrInvalid,
                        Node_srcloc(args));
  Node_next(args) = rest;
  return head;
}

void ExpandMacrosAndMacroLikeRecursively(Node fun, int nesting, IdGen* id_gen);

Node ExpandMacroBodyNodeRecursively(Node node, MacroContext* ctx) {
  // std::cout << "@@   Expand body node " << EnumToString(Node_kind(node))
  //          << "\n";
  // Note these may be written.
  std::map<Node, Node> dummy1;
  std::map<Node, Node> dummy2;
  switch (Node_kind(node)) {
    case NT::DefVar:
      if (NameIsMacro(Node_name(node))) {
        Node new_name = ctx->GetSymbol(Node_name(node));
        ASSERT(Node_kind(new_name) == NT::Id, "");
        Node type =
            ExpandMacroBodyNodeRecursively(Node_type_or_auto(node), ctx);
        Node initial = ExpandMacroBodyNodeRecursively(
            Node_initial_or_undef_or_auto(node), ctx);
        Node out = NodeNew(NT::DefVar);
        NodeInitDefVar(out, Node_name(new_name), type, initial,
                       Node_compressed_flags(node), kStrInvalid,
                       Node_srcloc(node));
        return out;
      }

      break;
    case NT::MacroId: {
      Node arg = ctx->GetSymbol(Node_name(node));

      Node replacement = NodeCloneRecursively(arg, &dummy1, &dummy2);
      // std::cout << "@@ expanding " << Node_name(node) << " "
      //          << EnumToString(Node_kind(arg)) << " " << arg.index() << " ->
      //          "
      //          << replacement.index() << "\n";

      if (Node_kind(replacement) == NT::EphemeralList) {
        Node args = Node_args(replacement);
        NodeFree(replacement);
        replacement = args;
      }

      return replacement;
    }
    case NT::MacroFor: {
      Name loop_var = Node_name(node);
      ASSERT(NameIsMacro(loop_var), "");
      Node args = ctx->GetSymbol(Node_name_list(node));
      ASSERT(Node_kind(args) == NT::EphemeralList, "");
      args = Node_args(node);
      NodeChain out;
      for (Node arg = args; !arg.isnull(); arg = Node_next(arg)) {
        ctx->SetSymbol(loop_var, arg);
        for (Node body = Node_body_for(node); !body.isnull();
             body = Node_next(body)) {
          Node exp = ExpandMacroBodyNodeRecursively(body, ctx);
          out.Append(exp);
        }
      }
      return out.First();
    }

    default:
      break;
  }

  Node clone = NodeCloneBasics(node);
  const NodeCore& core = gNodeCore[node];
  auto& core_clone = gNodeCore[clone];

  for (int i = 0; i < MAX_NODE_CHILDREN; ++i) {
    Node child = core.children_node[i];
    if (child.raw_kind() == kKindStr || child.raw_kind() == kKindName ||
        child.isnull()) {
      core_clone.children_node[i] = child;
      continue;
    }
    NodeChain new_children;

    do {
      Node exp = ExpandMacroBodyNodeRecursively(child, ctx);
      new_children.Append(exp);
      child = Node_next(child);
    } while (!child.isnull());

    core_clone.children_node[i] = new_children.First();
  }

  return clone;
}

constexpr int MAX_MACRO_NESTING = 8;

Node ExpandMacroInvocation(Node macro_invoke, int nesting, IdGen* id_gen) {
  // std::cout << "@@ Expand invoke of " << Node_name(macro_invoke)
  //           << " nesting=" << nesting << "\n";
  if (nesting >= MAX_MACRO_NESTING) {
    CompilerError(Node_srcloc(macro_invoke)) << "too many nested macros";
  }
  Node def_macro = Node_x_symbol(macro_invoke);
  if (Node_kind(def_macro) != NT::DefMacro) {
    CompilerError(Node_srcloc(macro_invoke)) << "not a macro";
  }
  Node params = Node_params_macro(def_macro);
  Node args = Node_args(macro_invoke);
  args = FixUpArgsForExprListRest(params, args);
  if (NodeNumSiblings(params) != NodeNumSiblings(args)) {
    CompilerError(Node_srcloc(macro_invoke))
        << "wrong number of macro arguments for " << Node_name(macro_invoke)
        << ": " << NodeNumSiblings(params) << " vs " << NodeNumSiblings(args);
  }

  MacroContext ctx(macro_invoke, id_gen, Node_srcloc(macro_invoke));

  for (Node p = params, a = args; !p.isnull();
       p = Node_next(p), a = Node_next(a)) {
    ctx.RegisterSymbolWithOwnership(Node_name(p), a);
  }

  // The ctx now "owns" the arg nodes so
  // we unlink them from each other and the invocation
  Node_args(macro_invoke) = kNodeInvalid;
  while (!args.isnull()) {
    Node next = Node_next(args);
    Node_next(args) = kNodeInvalid;
    args = next;
  }

  for (Node x = Node_gen_ids(def_macro); !x.isnull(); x = Node_next(x)) {
    ctx.GenerateNewSymbol(Node_name(x), Node_srcloc(macro_invoke));
  }

  NodeChain body_clone;
  for (Node body = Node_body_macro(def_macro); !body.isnull();
       body = Node_next(body)) {
    Node exp = ExpandMacroBodyNodeRecursively(body, &ctx);
    body_clone.Append(exp);
  }

  Node list = NodeNew(NT::EphemeralList);
  NodeInitEphemeralList(list, body_clone.First(), 0, kStrInvalid,
                        kSrcLocInvalid);
  ExpandMacrosAndMacroLikeRecursively(list, nesting + 1, id_gen);
  NodeFreeRecursively(macro_invoke);
  Node out = Node_args(list);
  NodeFree(list);
  return out;
}

void ExpandMacrosAndMacroLikeRecursively(Node fun, int nesting, IdGen* id_gen) {
  auto replacer = [nesting, id_gen](Node node, Node parent) -> Node {
    switch (Node_kind(node)) {
      case NT::MacroInvoke:
        return ExpandMacroInvocation(node, nesting, id_gen);
      case NT::ExprSrcLoc: {
        std::stringstream ss;
        ss << Node_srcloc(Node_expr(node));
        Node out = NodeNew(NT::ValString);
        NodeInitValString(out, StrNew(ss.str()), kStrInvalid,
                          Node_srcloc(node));
        NodeFreeRecursively(node);
        return out;
      }
      case NT::ExprStringify: {
        std::stringstream ss;
        ss << EnumToString(Node_kind(Node_expr(node)));
        Node out = NodeNew(NT::ValString);
        NodeInitValString(out, StrNew(ss.str()), kStrInvalid,
                          Node_srcloc(node));
        NodeFreeRecursively(node);
        return out;
      }

      default:
        return node;
    }
  };
  MaybeReplaceAstRecursivelyPost(fun, replacer, kNodeInvalid);
}

void ExpandMacrosAndMacroLike(const std::vector<Node>& mods) {
  for (Node mod : mods) {
    for (Node child = Node_body_mod(mod); !child.isnull();
         child = Node_next(child)) {
      if (Node_kind(child) == NT::DefFun) {
        IdGen idgen;
        ExpandMacrosAndMacroLikeRecursively(child, 0, &idgen);
      }
    }
  }

  for (Node mod : mods) {
    RemoveNodesOfType(mod, NT::DefMacro);
  }
}
}  // namespace cwerg::fe